models/flowers_oop.nim

```diff
@@ -0,0 +1,510 @@
+      initialize(self)
+    elif system.abs(self.translation.z) > 10:
+      initialize(self)
+    else:
+      self.velocity.x += rand(1.0) * 0.02 - 0.01
+      self.velocity.y += rand(1.0) * 0.02 - 0.01
+      self.velocity.z += rand(1.0) * 0.02 - 0.01
+
+const
+  OrbitSpeed = 0.5f
+  MinDistance = 5.0f
+  MaxDistance = 15.0f
+
+type
+  OrbitCamera = object
+    camera: Camera3D
+    angle: float32
+    distance: float32
+    target: Vector3
+
+proc initOrbitCamera(): OrbitCamera =
+  result = OrbitCamera(
+    camera: Camera3D(
+      position: Vector3(x: 2.0f, y: 2.0f, z: 2.0f),
+      target: Vector3(x: 0.0f, y: 0.0f, z: 0.0f),
+      up: Vector3(x: 0.0f, y: 1.0f, z: 0.0f),
+      fovy: 45.0f,
+      projection: Perspective
+    ),
+    angle: 0.0f,
+    distance: 10.0f,
+    target: Vector3(x: 0.0f, y: 0.0f, z: 0.0f)
+  )
+
+proc updateOrbitCamera(orbit: var OrbitCamera) =
+  # Handle mouse input for rotation
+  if isMouseButtonDown(MouseButton.Left):
+    let deltaX = getMouseDelta().x
+    orbit.angle -= deltaX * OrbitSpeed * getFrameTime()
+
+  # Handle mouse wheel for zoom
+  let wheel = getMouseWheelMove()
+  if wheel != 0.0f:
+    orbit.distance = clamp(orbit.distance - wheel, MinDistance, MaxDistance)
+
+  # Calculate new camera position
+  orbit.camera.position.x = orbit.target.x + sin(orbit.angle) * orbit.distance
+  orbit.camera.position.z = orbit.target.z + cos(orbit.angle) * orbit.distance
+  orbit.camera.position.y = orbit.target.y + orbit.distance * 0.5f
+
+proc main() =
+  # Initialization
+  # --------------------------------------------------------------------------------------
+  initWindow(screenWidth, screenHeight, "JSONverse shaders example - rhodonea")
+  defer: closeWindow()
+
+  var cube = genMeshCube(4.0f, 4.0f, 4.0f)
+  var skybox = loadModelFromMesh(cube)
+  var skyboxShader = loadShaderFromMemory(vertexSkyboxShader, fragmentSkyboxShader)
+  skybox.materials[0].shader = skyboxShader
+
+  var image = loadImage("resources/skybox.png")  # TODO Replace with your texture path
+  if image.data == nil:
+    raise newException(IOError, "Failed to load skybox.png texture")
+  var cubemap = loadTextureCubemap(image, CubemapLayout.AutoDetect)
+  reset(image)
+
+  # Debugging output
+  if skyboxShader.id == 0:
+    echo "Failed to load skyboxShader"
+    return
+
+  if cubemap.id == 0:
+    echo "Failed to load cubemap"
+    return
+
+  skybox.materials[0].maps[MaterialMapIndex.Cubemap].texture = cubemap
+
+  var img = genImageChecked(64, 64, 32, 32, DarkBrown, DarkGray)
+  var backgroundTexture = loadTextureFromImage(img)
+
+  let FLOWERS = 25
+
+  var flowers: seq[Flower]
+
+  flowers.setLen(FLOWERS + 1)
+
+  for f in 0..<flowers.len:
+    build(flowers[f])
+
+  defer:
+    unloadTexture(backgroundTexture.id)
+    unloadTexture(cubemap.id)
+    # unloadModel(skybox.id)
+    # unloadShader(skyboxShader.id)
+    # unloadImage(img.id)
+    for f in 0..<flowers.len:
+      `=destroy`(flowers[f])
```

Comment:
Destructors are run automatically. No need to call them explicitly. If you really need to (such as to free an image after loading it into a texture) and you don't depending on scopes, there's `reset`.